addons/mail/models/mail_thread.py

```diff
@@ -4430,17 +4430,32 @@ def _extract_partner_ids_for_notifications(self, message, msg_vals, recipients_d
 
     def _truncate_payload(self, payload):
         """
-        Check the payload limit of 4096 bytes to avoid 413 error return code.
+        Check the payload limit of 3990 bytes to avoid 413 error return code.
         If the payload is too big, we trunc the body value.
+        This limit has been chosen based on the extra bytes added by the encryption
+        using AES128GCM. Taking into account what are the extra bytes added at the header:
+        salt(16) + record_size(4) + key_len(1) + public_key (sender_public_key.len)
         :param dict payload: Current payload to trunc
         :return: The truncate payload;
         """
-        payload_length = len(str(payload).encode())
-        body = payload['options']['body']
+        payload_length = len(json.dumps(payload).encode())
+        # json.dumps defaults to translating unicode to hex codepoints (ensure_ascii=True)
+        # hence we need to check the length the body takes up in that format
+        # json string quotes are removed and the body is not encoded as it's already all ASCII
+        body = json.dumps(payload['options']['body'])[1:-1]
         body_length = len(body)
-        if payload_length > 4096:
-            body_max_length = 4096 - payload_length - body_length
-            payload['options']['body'] = body.encode()[:body_max_length].decode(errors="ignore")
+
+        if payload_length > 3990:
+            body_max_length = max(0, 3990 - payload_length + body_length)
+            # truncate to max length and try to loads again
+            # if there's any error, it will be a unicode error
+            # the error position gives us the start of the codepoint
+            # remove everything after that + the preceding escape marker (\u)
+            try:
+                truncated_body = json.loads(f'"{body[:body_max_length + 1]}"')
+            except json.decoder.JSONDecodeError as json_error:
+                truncated_body = json.loads(f'"{body[:json_error.pos-2]}"')
```

Comment:
@tde-banana-odoo / @awa-odoo 

Can you check if this part seems ok to you? It's a relatively important change vs what I got delegate for, though I'm confident it shouldn't pose any issues. (see second commit)

Reply:
Sorry, I've tried to figure out what we are trying to do here but I really don't understand why we use json loads to determine how you cut a body length, which is seemingly just a string.

Reply:
Edited the method doctring explaining the how and why, why the limit is what it is, and added a test where the limit is mocked that's hopefully easier to follow. In addition to the test that checks against the default limit.

Hopefully it's clearer for anyone that needs to dig around in there in the future. Or if nothing else I'll have acquired somewhat useless knowledge about encryption overhead :+1: